phalcon/DM/Pdo/Connection.zep

```diff
@@ -0,0 +1,154 @@
+ */
+
+namespace Phalcon\DM\Pdo;
+
+use InvalidArgumentException;
+use Phalcon\DM\Pdo\Connection\AbstractConnection;
+use Phalcon\DM\Pdo\Profiler\Profiler;
+use Phalcon\DM\Pdo\Profiler\ProfilerInterface;
+
+/**
+ * Provides array quoting, profiling, a new `perform()` method, new `fetch*()`
+ * methods
+ *
+ * @property array             $arguments
+ * @property PDO               $pdo
+ * @property ProfilerInterface $profiler
+ */
+class Connection extends AbstractConnection
+{
+    /**
+     * @var array
+     */
+    protected arguments = [];
+
+    /**
+     * Constructor.
+     *
+     * This overrides the parent so that it can take connection attributes as a
+     * constructor parameter, and set them after connection.
+     *
+     * @param string            $dsn
+     * @param string            $username
+     * @param string            $password
+     * @param array             $options
+     * @param array             $queries
+     * @param ProfilerInterface $profiler
+     */
+    public function __construct(
+        string dsn,
+        string username = null,
+        string password = null,
+        array options = [],
+        array queries = [],
+        <ProfilerInterface> profiler = null
+    ) {
+        var parts;
+        array available;
+
+        let parts     = explode(":", dsn),
+            available = [
+                "mysql"  : true,
+                "pgsql"  : true,
+                "sqlite" : true,
+                "mssql"  : true
+            ];
+
+        if !isset available[parts[0]] {
+            throw new InvalidArgumentException(
+                "Driver not supported [" . parts[0] . "]"
+            );
+        }
+
+
+        // if no error mode is specified, use exceptions
+        if !isset options[\PDO::ATTR_ERRMODE] {
+            let options[\PDO::ATTR_ERRMODE] = \PDO::ERRMODE_EXCEPTION;
+        }
+
+        // Arguments store
+        let this->arguments = [
+            dsn,
+            username,
+            password,
+            options,
+            queries
+        ];
+
+        // Create a new profiler if none has been passed
+        if profiler === null {
+            let profiler = new Profiler();
+        }
+        this->setProfiler(profiler);
+    }
+
+    /**
+     * The purpose of this method is to hide sensitive data from stack traces.
+     *
+     * @return array
+     */
+    public function __debugInfo() -> array
+    {
+        return [
+            "arguments" : [
+                this->arguments[0],
+                "****",
+                "****",
+                this->arguments[3],
+                this->arguments[4]
+            ]
+        ];
+    }
+
+    /**
+     * Connects to the database.
+     */
+    public function connect() -> void
+    {
+        var dsn, options, password, query, queries, username;
+
+        if !this->pdo {
+            // connect
+            this->profiler->start(__FUNCTION__);
```

Comment:
Can we make the profiler optional for perfomance?

Reply:
The `if` statement is in the Profiler class (`active`). It was a tradeoff between putting if statements in every `fetch` method and not having one method call vs. having the method call and the `if` is in the profiler.